### ISSUE TITLE
Is pervasive reference counting necessary?

### ISSUE THREAD

**Issue by ubolonton** [TheBB/libegit2#2]
The README mentions 2 reasons for reference counting. I understand the second, but not the first.

> Several user pointers may point to the same libgit2 structure

Shouldn't the library avoid wrapping the same libgit2 structure multiple times? The wrapping "user pointer" can be the owner, other code simply borrows the underlying libgit2 raw pointer, to do things with it, but never frees the memory itself. This way, the only type that needs reference counting is the `repo`.

Do you have an example that shows the necessity of reference counting due to the first reason?

**Comment by TheBB** [comment]
So you're suggesting the following (to see if I understand it correctly). When returning a git_whatever, instead of creating a new user pointer each time, we check whether one already exists for that pointer, and return that. Since only one such object is ever created per pointer, we can then rely on Emacs' garbage collector to tell us when an object is no longer needed. Good.

In order to keep such a persistent table of user pointers, we need to store [global references](https://phst.github.io/emacs-modules#global-references) to these objects. Otherwise the table becomes invalidated each time a new libgit-whatever function is called. But this also renders the Emacs GC useless: it will never decide to garbage collect global referenced objects, it has to wait for your module to free it with `free_global_ref`. Aren't we then back where we started?

**Comment by ubolonton** [comment]
> When returning a git_whatever, instead of creating a new user pointer each time, we check whether one already exists for that pointer, and return that

No, I'm suggesting putting `*git_whatever` in a user pointer without this check, since AFAICT, "one already exists for that pointer" cannot happen.

**Comment by TheBB** [comment]
Ah okay. Now that you mention it, that must be the case otherwise using libgit2 from C would already be all but impossible.

**Comment by TheBB** [comment]
https://github.com/TheBB/libegit2/pull/3
